baru_wallet/src/lib.rs

```diff
@@ -0,0 +1,652 @@
+use aes_gcm_siv::{
+    aead::{generic_array::GenericArray, Aead, NewAead},
+    Aes256GcmSiv,
+};
+use anyhow::bail;
+use anyhow::{Context, Result};
+use elements::sighash::SigHashCache;
+use elements::{
+    bitcoin::{
+        secp256k1::{PublicKey, SECP256K1},
+        Amount,
+    },
+    confidential,
+    secp256k1_zkp::SecretKey,
+    Address, AddressParams, AssetId, BlockHash, OutPoint, Transaction, TxOut, TxOutSecrets, Txid,
+};
+use elements::{
+    hashes::{hash160, Hash},
+    opcodes,
+    script::Builder,
+    secp256k1_zkp::{rand, Message},
+    SigHashType, TxIn,
+};
+use estimate_transaction_size::estimate_virtual_size;
+use hkdf::Hkdf;
+use itertools::Itertools;
+use rand::thread_rng;
+use rand::Rng;
+use serde::{Deserialize, Serialize};
+use sha2::Sha256;
+use std::ops::{Add, Sub};
+use std::str::FromStr;
+use std::{collections::HashMap, iter};
+
+pub use crate::client::Client;
+
+mod client;
+
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub enum Chain {
+    Elements,
+    Liquid,
+}
+
+impl From<Chain> for &AddressParams {
+    fn from(from: Chain) -> Self {
+        match from {
+            Chain::Elements => &AddressParams::ELEMENTS,
+            Chain::Liquid => &AddressParams::LIQUID,
+        }
+    }
+}
+
+impl FromStr for Chain {
+    type Err = WrongChain;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let lowercase = s.to_ascii_lowercase();
+        match lowercase.as_str() {
+            "elements" => Ok(Chain::Elements),
+            "liquid" => Ok(Chain::Liquid),
+            _ => Err(WrongChain(lowercase)),
+        }
+    }
+}
+
+#[derive(Debug, thiserror::Error)]
+#[error("Unsupported chain: {0}")]
+pub struct WrongChain(String);
+
+#[derive(Debug)]
+pub struct Wallet {
+    pub name: String,
+    encryption_key: [u8; 32],
+    secret_key: SecretKey,
+    pub sk_salt: [u8; 32],
+    chain: Chain,
+    cache: Vec<(Utxo, TxOut)>,
+}
+
+const SECRET_KEY_ENCRYPTION_NONCE: &[u8; 12] = b"SECRET_KEY!!";
+
+impl Wallet {
+    pub async fn sync(&mut self, client: &impl Client) -> Result<()> {
+        self.cache = client.get_txouts(self.address()).await?;
+        Ok(())
+    }
+    pub fn utxos(&self) -> Vec<(Utxo, TxOut)> {
+        self.cache.clone()
+    }
```

Comment:
I don't think we should expose our UTXOs. That reveals too much of the inner structure of this type.